query-coordinator/src/main/scala/com/socrata/querycoordinator/rollups/QueryRewriter.scala

```diff
@@ -0,0 +1,75 @@
+package com.socrata.querycoordinator.rollups
+
+import com.socrata.querycoordinator.rollups.QueryRewriter.{Anal, Expr, RollupName}
+import com.socrata.querycoordinator.{Schema, SchemaWithFieldName}
+import com.socrata.soql.environment.{ColumnName, TableName}
+import com.socrata.soql.functions.{SoQLFunctions}
+import com.socrata.soql.{BinaryTree, SoQLAnalysis}
+import com.socrata.soql.types.{SoQLType}
+
+/**
+  *
+  * This is the interface of the QueryRewriter
+  *
+  * Any class wishing to use the QueryRewriter must use the interface, not a concrete implementation.
+  */
+abstract class QueryRewriter {
+
+  /**
+    * *Map rollup definitions into query analysis in schema context
+    *
+    * ToDo: Should this return merged analysis?
+    *
+    */
+  def analyzeRollups(schema: Schema, rollups: Seq[RollupInfo],
+                     getSchemaByTableName: TableName => SchemaWithFieldName): Map[RollupName, BinaryTree[Anal]]
+
+  /**
+    * Creates possible rewrites for the query given the rollups
+    *
+    */
+  def possibleRewrites(q: Anal, rollups: Map[RollupName, Anal]): Map[RollupName, Anal]
+  def possibleRewrites(q: BinaryTree[Anal], rollups: Map[RollupName, BinaryTree[Anal]], requireCompoundRollupHint: Boolean): (BinaryTree[Anal], Seq[String])
+  def possibleRewrites(q: Anal, rollups: Map[RollupName, Anal], debug: Boolean): Map[RollupName, Anal]
+
+  /**
+    * Returns best scored rollup
+    *
+    */
+  def bestRollup(rollups: Seq[(RollupName, Anal)]): Option[(RollupName, Anal)]
+
+}
+
+object QueryRewriter {
+  import com.socrata.soql.typed._ // scalastyle:ignore import.grouping
+
+  type Anal = SoQLAnalysis[ColumnId, SoQLType]
```

Comment:
Nit: I'd prefer renaming this type alias to `Analysis`. 🤣

Reply:
100%. It's already there, and I'm surprised it made it to the code. That would be a large change as it's used in many places, so I'll make a separate PR